### ISSUE TITLE
Running failed 

### ISSUE THREAD

**Issue by bkovalev** [yahoo/tensorflow#5]
Hi,

I have 1 PS and 2 Workers with ConnectX 4 - 100Gb IPoIB
Tensorflow 1.0.0, CUDA 8.0 , Ubuntu 16.04, MLNX_OFED_LINUX-4.0-0.1.5.0-ubuntu16.04-x86_64
**PS starting ok**.
bazel-bin/inception/imagenet_vgg16_distributed_train_rdma --batch_size=32 --protocol=grpc_rdma --job_name=ps --ps_hosts='12.12.12.30:2222' --worker_hosts='12.12.12.22:2222,12.12.12.24:2222' --task_id=0

 tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:200] Initialize GrpcChannelCache for job ps -> {0 -> localhost:2222}
I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:200] Initialize GrpcChannelCache for job worker -> {0 -> 12.12.12.22:2222, 1 -> 12.12.12.24:2222}
I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:241] Started server with target: grpc://localhost:2222
I tensorflow/core/distributed_runtime/rdma/rdma_mgr.cc:38] connecting to remote node /job:worker/replica:0/task:1
I tensorflow/core/distributed_runtime/rdma/rdma_mgr.cc:38] connecting to remote node /job:worker/replica:0/task:0

**Worker (task_id 1) - waiting to second Worker** 
bazel-bin/inception/imagenet_vgg16_distributed_train_rdma --batch_size=16 --job_name=worker --protocol=grpc_rdma --ps_hosts='12.12.12.30:2222' --worker_hosts='12.12.12.22:2222,12.12.12.24:2222' --task_id=1 --data_dir=/data/imagenet_data/ --train_dir=/data/imagenet_train/

I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:241] Started server with target: grpc://localhost:2222
I tensorflow/core/distributed_runtime/rdma/rdma_mgr.cc:38] connecting to remote node /job:worker/replica:0/task:0

**Worker(task_id 0) - failed with following error:**
 
I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:200] Initialize GrpcChannelCache for job ps -> {0 -> 12.12.12.30:2222}                                                                                                    
I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:200] Initialize GrpcChannelCache for job worker -> {0 -> localhost:2222, 1 -> 12.12.12.24:2222}                                                                           
I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:241] Started server with target: grpc://localhost:2222                                                                                                                 
I tensorflow/core/distributed_runtime/rdma/rdma_mgr.cc:38] connecting to remote node /job:worker/replica:0/task:1 
I tensorflow/core/distributed_runtime/rdma/rdma_mgr.cc:38] connecting to remote node /job:ps/replica:0/task:0     
INFO:tensorflow:Summary name CrossEntropyLoss/value (raw) is illegal; using CrossEntropyLoss/value__raw_ instead. 
INFO:tensorflow:Summary name conv0/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv0/weights/Regularizer/L2Regularizer/value__raw_ instead.                                                                     
INFO:tensorflow:Summary name conv1/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv1/weights/Regularizer/L2Regularizer/value__raw_ instead.                                                                     
INFO:tensorflow:Summary name conv2/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv2/weights/Regularizer/L2Regularizer/value__raw_ instead.                                                                     
INFO:tensorflow:Summary name conv3/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv3/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name conv4/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv4/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name conv5/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv5/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name conv6/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv6/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name conv7/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv7/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name conv8/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv8/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name conv9/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv9/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name conv10/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv10/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name conv11/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv11/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name conv12/weights/Regularizer/L2Regularizer/value (raw) is illegal; using conv12/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name fc1/weights/Regularizer/L2Regularizer/value (raw) is illegal; using fc1/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name fc2/weights/Regularizer/L2Regularizer/value (raw) is illegal; using fc2/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name logits/weights/Regularizer/L2Regularizer/value (raw) is illegal; using logits/weights/Regularizer/L2Regularizer/value__raw_ instead.
INFO:tensorflow:Summary name total_loss (raw) is illegal; using total_loss__raw_ instead.
INFO:tensorflow:SyncReplicasV2: replicas_to_aggregate=2; total_num_replicas=2
INFO:tensorflow:2017-03-29 12:43:45.868750 Supervisor
I tensorflow/core/distributed_runtime/master_session.cc:1012] Start master session 1199eb27f7f89798 with config:
allow_soft_placement: true

F tensorflow/core/distributed_runtime/rdma/rdma.cc:116] Check failed: wc_[i].status == IBV_WC_SUCCESS Failed status
transport retry counter exceeded 12 58961952 129
Aborted (core dumped)

IB was checked and works good.
Any ideas? Maybe a firmware version? API? 

[imagenet_vgg16_distributed_train_rdma.txt](https://github.com/yahoo/tensorflow/files/878565/imagenet_vgg16_distributed_train_rdma.txt)
[vgg16_distributed_train_rdma.txt](https://github.com/yahoo/tensorflow/files/878564/vgg16_distributed_train_rdma.txt)
[BUILD.txt](https://github.com/yahoo/tensorflow/files/878563/BUILD.txt)

Thanks

**Comment by junshi15** [comment]
Is your infiniband configured in "native" RDMA mode or IPoIB mode? I have not tried IPoIB, don't know ibverbs API supports IPoIB or not.

**Comment by bkovalev** [comment]
Thanks junshi15!
Good jobs!
My infiniband configured in IPoIB.

**Comment by bkovalev** [comment]
Hi junshi15,
I am running on "native" RDMA mode and have the same issue.
can you provide me how to run it correct.

Thanks

**Comment by junshi15** [comment]
The way you launch your job seems to be correct. At least you have got to a line in rdma.cc, this means the code is running in the rdma path. 

The error message suggests there is some problem in the network. I am afraid I won't be of any help with your network setting. But this page may be helpful.
http://www.rdmamojo.com/2013/02/15/ibv_poll_cq/
You can search for "Transport Retry Counter Exceeded" in that page.

The error code says:
IBV_WC_RETRY_EXC_ERR (12) - Transport Retry Counter Exceeded: The local transport timeout retry counter was exceeded while trying to send this message. This means that the remote side didn't send any Ack or Nack. If this happens when sending the first message, usually this mean that the connection attributes are wrong or the remote side isn't in a state that it can respond to messages. If this happens after sending the first message, usually it means that the remote QP isn't available anymore. Relevant for RC QPs.

Then, in the comment section, one user (Parthiban) experienced the same error. In his case, he had more than one network interface, but the first one found by the program was not the one in the subnet. It may be useful reading the Q/A section of that blog.

**Comment by junshi15** [comment]
In your log (for all three nodes), do you ever see the line below:

I tensorflow/contrib/verbs/rdma.cc:502] channel already connected

This should take place at the beginning of the log file. If you don't see this line, chances are the connection among nodes was not established.

**Comment by bkovalev** [comment]
Ok, now I understand my issue. 
Because I have ConnectX 4, I have two devices. And first from them in Down.
In your rdma.cc  it's running on first from list (which is Down). 
Sure it's not will work.

As improvement, I want to put a Device as a parameter in TF command and add i the rdm.cc a check if the device in Active status.
 
Thanks a lot for your hard work.

**Comment by junshi15** [comment]
Cool. Glad you found the cause.

Yes, you need to parameterize the following line:
https://github.com/yahoo/tensorflow/blob/jun_r1.0/tensorflow/core/distributed_runtime/rdma/rdma.cc#L59

BTW, we have opened a PR at google tensorflow. Hopefully this piece of code gets merged into official tensorflow repo. 
https://github.com/tensorflow/tensorflow/pull/8943

**Comment by fanlu** [comment]
@bkovalev I found the same issue, Could you share your changed code to me?thx
